extensions/cl_ext_image_tiling_control.asciidoc

```diff
@@ -0,0 +1,214 @@
+// Creative Commons Attribution 4.0 International License; see
+// http://creativecommons.org/licenses/by/4.0/
+
+:data-uri:
+:icons: font
+include::../config/attribs.txt[]
+include::{generated}/api/api-dictionary-no-links.asciidoc[]
+:source-highlighter: coderay
+
+= cl_ext_image_tiling_control
+
+== Name Strings
+
+`cl_ext_image_tiling_control`
+
+== Contact
+
+Kevin Petit (kevin.petit 'at' arm.com)
+
+== Contributors
+
+Kevin Petit, Arm Ltd. +
+Plamen Petkov, Arm Ltd. +
+
+== Notice
+
+Copyright (c) 2021 Arm Ltd.
+
+== Status
+
+Draft
+
+== Version
+
+Built On: {docdate} +
+Version: 0.1.0
+
+== Dependencies
+
+This extension is written against the OpenCL Specification version 3.0.10.
+
+This extension requires OpenCL 3.0.
+
+== Overview
+
+This extension gives applications explicit control over how image data is laid out in memory.
+
+== New API Types
+
+[source,c]
+----
+typedef cl_uint cl_image_tiling_ext;
+
+#define CL_IMAGE_TILING_LINEAR_EXT   1
+#define CL_IMAGE_TILING_OPTIMAL_EXT  2
+
+typedef cl_bitfield cl_device_image_tiling_capabilities_ext;
+
+#define CL_DEVICE_IMAGE_TILING_DEVICE_ACCESS_EXT (1 << 0)
+#define CL_DEVICE_IMAGE_TILING_HOST_ACCESS_EXT   (1 << 1)
+----
+
+== New API Enums
+
+Accepted value for the _param_name_ parameter to *clGetDeviceInfo*:
+
+[source,c]
+----
+CL_DEVICE_IMAGE_TILING_CAPABILITIES_EXT 0x4234
+----
+
+Accepted value for the _properties_ parameter to *clCreateImageWithProperties*:
+
+[source,c]
+----
+CL_MEM_IMAGE_TILING_EXT       0x4235
+----
+
+Accepted value for the _param_name_ parameter to *clGetImageInfo*:
+
+[source,c]
+----
+CL_IMAGE_TILING_EXT 0x4236
+----
+
+== Modifications to the OpenCL API Specification
+
+(Modify Section 4.2, *Querying Devices*) ::
++
+--
+
+(Add the following to Table 5, _List of supported param_names by *clGetDeviceInfo*_) ::
++
+--
+
+[cols="1,1,4",options="header"]
+|====
+| cl_device_info
+| Return Type
+| Description
+
+| {CL_DEVICE_IMAGE_TILING_CAPABILITIES_EXT}
+| {cl_device_image_tiling_capabilities_ext_TYPE}
+| Returns the image tiling capabilities supported by the device.
+ +
+- {CL_DEVICE_IMAGE_TILING_DEVICE_ACCESS_EXT} is always set indicating that all
+  implementations that support `cl_ext_image_tiling_control` must support
+  device access to tiled images. +
+- {CL_DEVICE_IMAGE_TILING_HOST_ACCESS_EXT} is set when it is allowed to pass
+    {CL_MEM_IMAGE_TILING_EXT} with a value different from {CL_IMAGE_TILING_LINEAR_EXT}
+    in the _properties_ given to *clCreateImageWithProperties* when creating host-accessible images.
+    Host-accessible images are all images created without {CL_MEM_HOST_NO_ACCESS}.
+|====
+
+--
+--
+
+(Modify section 5.3.1, *Creating Image Objects*) ::
++
+--
+{CL_MEM_IMAGE_TILING_EXT} can be passed as part of the _properties_ parameter
+to *clCreateImageWithProperties* to control the tiling used for the image being
+created. The following values are accepted:
+
+  * {CL_IMAGE_TILING_LINEAR_EXT}, which is the default value used if
```

Comment:
Yes, I agree, and I had a similar comment above.

One possible way to solve this problem is to have three "tiling" states:

1. An explicitly linear image.
2. An explicitly tiled image.
3. A "default" or "unspecified" layout where the implementation can choose whether the image is linear or tiled or something else.

(1) and (2) are already defined by this extension.  (3) would need to be added.  If we added (3) and made it the default value I think it would retain backwards compatibility and avoid any performance penalties?

Reply:
I think both having a separate default or optimal covering non-linear as well as linear tiling modes should work.
Default may be unnecessary if optimal allows both.
In case of default or optimal tiling mode, if implementation is free to choose optimal or linear mode, what should clGetImageInfo CL_IMAGE_TILING_EXT query return? Should it return default/optimal tiling/layout or should it return optimal / linear as chosen by implementation?